### PR TITLE
[2021.2] [ShaderGraph] Fix property reference names to not use GUIDs

### PR DESCRIPTION
**Pull request by cdxntchou**
### Purpose of this PR

Fix for bugs:
https://fogbugz.unity3d.com/f/cases/1306820/
https://fogbugz.unity3d.com/f/cases/1300484/

When adding properties to the shadergraph blackboard, they are given "reference" names (names used in the HLSL code, and as identifiers for manipulating them in C# script) that include a GUID, which looks like a large string of random numbers to the user:

image

As a result, these names are not very user friendly, and we've received a lot of complaints about constantly having to change the reference names to something simpler and more readable.

With this PR, any newly created properties will instead use the new reference naming scheme, which keeps the reference name as close to the display name as possible, while maintaining it as a unique and valid HLSL identifier:

image

For any pre-existing properties, users can opt to upgrade the default naming scheme, via right click menu on the reference label:
image

This PR also adds the "Reset Reference" option (that removes overrides) to the same right click menu.
It also adds a "bold" style to the label when overridden:
image

Previously bool keywords only showed up in the Material inspector if they were using a ref name that ended in _ON.  This meant they would not show up when duplicated, or when using an override name:
image

With this PR all of the bool keywords show up in the material and are controllable.  The ref names are unchanged, but can now be overridden or use the new naming scheme, and will work either way.
image

Enum-based keywords always worked:
image

And continue to work:
image

As part of this PR, the paths that copy shader inputs to a graph were unified to use a single function (AddCopyOfShaderInput), simplifying the code and ensuring we always run the same sanitization and de-duplication on all paths.  As part of this, I did a bit of standardization to the ways shader inputs were copied, moving handling of AbstractShaderProperty fields from the virtual Copy() function to AddCopyOfShaderInput.  I double checked that all fields were copied appropriately, fixing a few issues (https://docs.google.com/spreadsheets/d/1wwLkaAxD9oeXfYUFIuUZr5F9tHhK3ylRanHACGY-Yxo/edit#gid=0)

---
### Testing status

Added new unit tests to test name deduplication, name sanitization, and default reference name behavior.

- [x] Tested that pre-existing properties and keywords (of all types) maintain their pre-existing reference name, whether using an override or the default name.
- [x] Tested that pre-existing properties can be reset to default name, and upgraded to use the new default naming scheme, via the right click menus on the ref name label.
- [x] Tested that newly created properties are created using the new default naming scheme, and can be overridden and reverted to default via the right click menus.
- [x] Tested creating new properties via the blackboard "+" menu with different names, duplicate names, default names.  Always resulted in properties with non-duplicate display names, and non-duplicate reference names.
- [x] Tested ConvertTo>Property creates a property with a non-conflicting name, and the names show up correctly in all locations.
- [x] Tested modifying display names of existing properties, attempting to get duplicate reference names via different means (undo/redo, ref name overrides on some properties and not others, resetting overrides) -- no detected way to get duplicate reference names.
- [x] Tested copy/pasting properties from other graphs, with unique or colliding names, using default or override on the ref name.  Ref names are not changed, as long as there are no collisions.  However, the copy may be changed from a default ref name to using an override ref name, even when there are no collisions.  A property using an old default ref name (w GUID) is changed to use the new default, but with an override to match the old name.  Note that copies are upgraded to use the new default naming scheme (overrides are used to match old names).
- [x] Tested drag/drop properties from other graphs, with unique or colliding names, using default or override on the ref name.  No issues with duplicate ref names in the results.  Note that copies are upgraded to use the new default naming scheme (overrides are used to match old names).
- [x] Tested duplicating properties on the blackboard -- names are made unique.  Note that copies are upgraded to use the new default naming scheme.
- [x] Tested ConvertToProperty on float node, vector4 node -- result is correct and no duplicate naming.
- [x] Tested Keyword reference names pass all of the same tests as properties above.
- [x] Tested that modifying display names can undo/redo correctly, including the default reference name change.
- [x] Tested that modifying reference name (override) can undo/redo correctly
- [x] Tested that resetting the reference name (removing override) can undo/redo correctly
- [x] Tested that upgrading the a pre-existing property's default reference name (using new naming convention) can undo/redo correctly.
- [x] Tested that new default property reference names are reflected in the generated shaders correctly, and work.
- [x] Verified that new default property reference names are serialized to disk correctly.
- [x] Tested that new default named keywords are reflected in the generated shaders correctly, and work.
- [x] Verified that new default keyword reference names are serialized to disk correctly.
- [x] Test Reset Reference on blackboard property
- [x] Test blackboard property scrolling moving the containing scroll view
- [x] Tested that renaming a property in any of the three locations will properly update all locations.  (Blackboard, Node Settings Graph Inspector for node or for property)
- [x] Tested SamplerState property types, which always use a hard-coded reference name based on their settings -- disallows changing ref name, no reset available.
- [x] Tested using shaderlab keywords as property names, all are renamed to avoid conflicts
- [x] Tested colliding shader keyword names and property names
 
Discovered Issues:
- [ ] Changing VirtualTexture property override ref names on old ShaderGraphs generates an error from the preview Material setup code (just once).  This appears to be a pre-existing issue though, it happens in master as well.  Will open a bug for it.
- [ ] Setting a Keyword to or from the "Predefined" definition does not hide or unhide the "local/global" enum, until you re-select it.  Pre-existing issue.
- [ ] Toggline exposed on a shader property with the NODE selected does not update the exposed icon on the blackboard view.  Pre-existing issue.

ShaderGraph PR job:  🟢 
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252Fproperty-guids3/.yamato%252Fall-shadergraph.yml%2523PR_ShaderGraph_trunk/5084514/job/pipeline

HDRP PR job:  matches failures in master
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252Fproperty-guids3/.yamato%252Fall-hdrp.yml%2523PR_HDRP_trunk/5084544/job/pipeline

Universal PR job:  matches failures in master
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252Fproperty-guids3/.yamato%252Fall-universal.yml%2523PR_Universal_trunk/5087949/job/pipeline

Corresponding Master ABV:  eba6d6a
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252F_abv.yml%2523all_project_ci_nightly_trunk/5068250/job

Failing tasks:
HDRP on Linux_Vulkan_playmode_cache_mono_Linear on version trunk: failing same 4 tests same way
HDRP on OSX_Metal_playmode_cache_mono_Linear on version trunk: failing same 8 tests same way
HDRP on Win_Vulkan_playmode_cache_mono_Linear on version trunk: failing same 4 tests same way
HDRP on Win_DX12_playmode_cache_mono_Linear on version trunk: failing same 5 tests same way
HDRP on Win_DX11_playmode_XR_cache_mono_Linear on version trunk: both failed script compile same way
HDRP on Win_DX11_playmode_cache_mono_Linear on version trunk: both failed same 4 tests same way
Universal on Win_DX11_playmode_XR_cache_mono_Linear on version trunk: both failed script compile same way
Build Universal_Stereo on Win__mono_Linear_Standalone_cache_build_Player on version trunk: both failed with error "On demand scheduler requested to import while stopped"
Universal on Android_OpenGLES3_Standalone_cache_il2cpp_Linear on version trunk: both failed same 3 tests same way
Universal on Android_Vulkan_Standalone_cache_il2cpp_Linear on version trunk: both failed same 3 tests same way
Universal on iPhone_Metal_Standalone_cache_il2cpp_Linear on version trunk: both failed same 1 test same way

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Review by LandonTownsendUnity — APPROVED** [review]
Developer testing is sufficient. Only thing I wanted to make sure of was that you could enable the URP Lit features attached to keywords ending in _OFF; it turns out you can, and actually expose those keywords, which I'm very happy with.
shadergraph_off_keywords
Approving.

**Review by LandonTownsendUnity — CHANGES_REQUESTED** [review]
Actually, now that all keywords are added to the material inspector regardless of _ON suffix, there is now no way to choose not to expose a keyword to the material inspector. That should probably be added.
Additionally, now boolean keywords marked as "predefined" are added to the material inspector. This should not happen, as they can not be used like other keywords, and toggling them in the material inspector does nothing.

**Comment by LandonTownsendUnity** [comment]
Tested to validate that predefined keywords still work correctly, in both the build and the editor

**Review by alindmanUnity — APPROVED** [review]
UX sign off

**Comment by cdxntchou** [comment]
> 
> 
> Actually, now that all keywords are added to the material inspector regardless of _ON suffix, there is now no way to choose not to expose a keyword to the material inspector. That should probably be added.
> Additionally, now boolean keywords marked as "predefined" are added to the material inspector. This should not happen, as they can not be used like other keywords, and toggling them in the material inspector does nothing.

Pushed a fix -- now keywords allow the user to control their exposed state, and will version old keywords based on the presence of the _ON suffix to maintain old behavior.   Predefined and Built-in keywords are forced to be not exposed.

**Review by Nightmask3 — APPROVED** [review]
This looks good! 

In particular I approve of the rigorous testing and listing of test cases, really good PR 👍 

Also like the ability to upgrade from old reference name formats to the new ones, and the explicit functions added for sanitized display name/reference name changes vs. non-sanitized versions, makes it a lot more clear what your options are as a coder working with that data and what's expected.

**Comment by Nightmask3** [comment]
> This looks good!
> 
> In particular I approve of the rigorous testing and listing of test cases, really good PR 👍
> 
> Also like the ability to upgrade from old reference name formats to the new ones, and the explicit functions added for sanitized display name/reference name changes vs. non-sanitized versions, makes it a lot more clear what your options are as a coder working with that data and what's expected.

And of course, appreciate the tremendous workflow improvements of not having to rename your auto-generated reference names everytime, that being the whole point of the PR🤣

**Review by LandonTownsendUnity — APPROVED** [review]
Now that boolean keywords can be exposed by choice, and keywords upgrade properly to be exposed or unexposed depending on whether they end in _ON, and predefined keywords are locked to unexposed, I think this PR is good.

File: com.unity.shadergraph/Editor/Data/Implementation/NodeUtils.cs [hunk 1]
```diff
@@ -581,6 +581,120 @@ public static string GetHLSLSafeName(string input)
...
+            "oneminussrccolor",
+            "dstalpha",
+            "oneminusdstalpha",
+            "srcalphasaturate",
+            "oneminussrcalpha",
+            "constantcolor",
+            "oneminusconstantcolor",
+            "constantalpha",
+            "oneminusconstantalpha",
+        };
```
Comment (esmelusina) [inline com.unity.shadergraph/Editor/Data/Implementation/NodeUtils.cs:696]: Anyway ShaderLab can expose these?

**Review by esmelusina — APPROVED** [review]

**Comment by ValGrimm-U3D** [comment]
Related documentation ticket: https://unity3d.atlassian.net/browse/DOC-2823

File: com.unity.shadergraph/Editor/Data/Implementation/NodeUtils.cs (hunk 1 above)
Comment (jessebarker) [reply com.unity.shadergraph/Editor/Data/Implementation/NodeUtils.cs:696]: Seems like we could file a bug to get this kind of thing exposed (nothing to hold this PR up for, just something that would make things a bit cleaner in the future).

**Comment by cdxntchou** [comment]
Note: backporting to 21.1 would first require backporting #3085

**Pull request merged**